Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -143,35 +143,50 @@ open class BasicFormat: SyntaxRewriter {
     return node.requiresIndent
   }
 
-  /// Whether a leading newline on `token` should be added.
-  open func requiresLeadingNewline(_ token: TokenSyntax) -> Bool {
-    // We don't want to add newlines inside string interpolation
-    if isInsideStringInterpolation(token) {
+  open func requiresNewline(between first: TokenSyntax?, and second: TokenSyntax?) -> Bool {
+    // We don't want to add newlines inside string interpolation.
+    // When first or second `TokenSyntax` is a multiline quote we want special handling
+    // event it's inside a string interpolation, because it still requires newline
+    // after open quote and before close quote.
+    if let first,
+      first.tokenKind != .multilineStringQuote,
+      second?.tokenKind != .multilineStringQuote,
+      isInsideStringInterpolation(first)
+    {
       return false
-    }
-
-    if token.requiresLeadingNewline {
-      return true
-    }
-
-    var ancestor: Syntax = Syntax(token)
-    while let parent = ancestor.parent {
-      ancestor = parent
-      if ancestor.position != token.position || ancestor.firstToken(viewMode: viewMode) != token {
-        break
-      }
-      if let ancestorsParent = ancestor.parent, childrenSeparatedByNewline(ancestorsParent) {
+    } else if let second {
+      if second.requiresLeadingNewline {
```

Comment:
Definitely not part of this PR, but eventually I would like to remove `requiresLeadingNewline` from CodeGen altogether and just list the tokens that require a leading newline in this function. Same for `childrenSeparatedByNewline`. I don’t think we necessarily need to CodeGen it, it could just be implemented manually in this file.